cmd/nitro/init.go

```diff
@@ -173,7 +186,221 @@ func validateBlockChain(blockChain *core.BlockChain, expectedChainId *big.Int) e
+		header = rawdb.ReadHeader(r.chainDb, header.ParentHash, num-1)
+	}
+	height := header.Number.Uint64()
+	for len(r.heights) > 0 && r.heights[len(r.heights)-1] > height {
+		if overwrite {
+			r.roots = r.roots[:len(r.roots)-1]
+			r.heights = r.heights[:len(r.heights)-1]
+		} else {
+			return nil
+		}
+	}
+	if len(r.heights) > 0 && r.heights[len(r.heights)-1]+minRootDistance > height {
+		return nil
+	}
+	r.roots = append(r.roots, header.Root)
+	r.heights = append(r.heights, height)
+	return nil
+}
+
+var hashListRegex = regexp.MustCompile("^(0x)?[0-9a-fA-F]{64}(,(0x)?[0-9a-fA-F]{64})*$")
+
+// Finds important roots to retain while proving
+func findImportantRoots(ctx context.Context, chainDb ethdb.Database, stack *node.Node, nodeConfig *NodeConfig, cacheConfig *core.CacheConfig, l1Client arbutil.L1Interface, rollupAddrs arbnode.RollupAddresses) ([]common.Hash, error) {
+	initConfig := &nodeConfig.Init
+	chainConfig := arbnode.TryReadStoredChainConfig(chainDb)
+	if chainConfig == nil {
+		return nil, errors.New("database doesn't have a chain config (was this node initialized?)")
+	}
+	arbDb, err := stack.OpenDatabase("arbitrumdata", 0, 0, "", true)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		err := arbDb.Close()
+		if err != nil {
+			log.Warn("failed to close arbitrum database after finding pruning targets", "err", err)
+		}
+	}()
+	roots := importantRoots{
+		chainDb: chainDb,
+	}
+	genesisNum := chainConfig.ArbitrumChainParams.GenesisBlockNum
+	genesisHash := rawdb.ReadCanonicalHash(chainDb, genesisNum)
+	genesisHeader := rawdb.ReadHeader(chainDb, genesisHash, genesisNum)
+	if genesisHeader == nil {
+		return nil, errors.New("missing L2 genesis block header")
+	}
+	err = roots.addHeader(genesisHeader, false)
+	if err != nil {
+		return nil, err
+	}
+	if initConfig.Prune == "validator" {
+		if l1Client == nil {
+			return nil, errors.New("an L1 connection is required for validator pruning")
+		}
+		callOpts := bind.CallOpts{
+			Context:     ctx,
+			BlockNumber: big.NewInt(int64(rpc.FinalizedBlockNumber)),
+		}
+		rollup, err := staker.NewRollupWatcher(rollupAddrs.Rollup, l1Client, callOpts)
+		if err != nil {
+			return nil, err
+		}
+		latestConfirmedNum, err := rollup.LatestConfirmed(&callOpts)
+		if err != nil {
+			return nil, err
+		}
+		latestConfirmedNode, err := rollup.LookupNode(ctx, latestConfirmedNum)
+		if err != nil {
+			return nil, err
+		}
+		confirmedHash := latestConfirmedNode.Assertion.AfterState.GlobalState.BlockHash
+		confirmedNumber := rawdb.ReadHeaderNumber(chainDb, confirmedHash)
+		var confirmedHeader *types.Header
+		if confirmedNumber != nil {
+			confirmedHeader = rawdb.ReadHeader(chainDb, confirmedHash, *confirmedNumber)
+		}
+		if confirmedHeader != nil {
+			err = roots.addHeader(confirmedHeader, false)
+			if err != nil {
+				return nil, err
+			}
+
+			validatorDb := rawdb.NewTable(arbDb, arbnode.BlockValidatorPrefix)
+			lastValidated, err := staker.ReadLastValidatedFromDb(validatorDb)
```

Comment:
why is that inside if confirmeHeader != nil?
If not erroring-out in this case, you should still try to not prune latestValidatedFrom Db

Reply:
My thinking was that if we don't have the latest confirmed block, then the node needs to catch up before it validates. But I guess people might be re-validating the entire chain, and the safest option anyways is to keep everything we might need. I'll change this :+1: